File: javascript/backend.js
var autobahn = require('autobahn');

var connection = new autobahn.Connection({
    url: 'ws://yycjs.meejah.ca/ws',
    realm: 'demo'
});

var _clients = new Map();
function* _color_generator() {
    while (true) {
	yield {r: 255, g: 0, b: 0}
	yield {r: 0, g: 255, b: 0}
	yield {r: 0, g: 0, b: 255}
    }
}
var _client_color = _color_generator();

function client_create(client_id) {
    // each client gets a 16x16 array of bools
    var new_board = new Array(16);
    for (var i=0; i < 16; i++) {
	new_board[i] = new Array(16);
	for (var j=0; j < 16; j++) {
	    new_board[i][j] = false;
	}
    }
    // assign a color to the client
    var new_color = _client_color.next().value;
    // freshly-minted client
    return {
	id: client_id,
	color: new_color,
	board: new_board,
    }
}

function board_click(client_id, x, y) {
    // for each click on a square, we just toggle the state (on/off)
    client = _clients.get(client_id);
    console.log('board_click', client_id, x, y, client.board[x][y]);
    client.board[x][y] = !client.board[x][y];
}

function board_color_at(x, y) {
    // union of all clients' state at a particular location
    var color = [0, 0, 0];

    _clients.forEach(function (client) {
	if (client.board[x][y]) {
	    // taking max of each component, not e.g. average
	    color = [
		Math.max(color[0], client.color.r),
		Math.max(color[1], client.color.g),
		Math.max(color[2], client.color.b),
	    ];
	}
    });
    return color;
}

//board_click(1, 0, 0);
//console.log(board_color_at(0, 0));

connection.onopen = function (session) {
    console.log('joined');

    // an API method, exposed to clients as "meejah.click"
    function client_click(args, kwargs, details) {
	var x = args[0];
	var y = args[1];

	board_click(details.caller, x, y);
	console.log('publish meejah.board_update', [x, y, board_color_at(x, y)]);
	session.publish('meejah.board_update', [x, y, board_color_at(x, y)]);
    }
    session.register('meejah.click', client_click);

    // API method to get entire board-state (e.g. at client startup)
    function board_state() {
	console.log("board_state");
	var board = new Array(16);
	for (var x=0; x < 16; x++) {
	    board[x] = new Array(16);
	    for (var y=0; y < 16; y++) {
		board[x][y] = board_color_at(x, y);
	    }
	}
	return board;
    }
    session.register('meejah.get_board', board_state);

    /* notice when clients come and go */
    function client_left(args) {
	var session_id = args[0];
        _clients.delete(session_id);
	session.publish("meejah.board_invalid");
    }
    function client_join(args) {
	var details = args[0];
	_clients.set(details.session, client_create(details.session));
        session.publish("meejah.board_invalid")	
    }
    session.subscribe("wamp.session.on_join", client_join)
    session.subscribe("wamp.session.on_leave", client_left)

    // API method to get client-count
    session.register(
	'meejah.get_client_count',
	function() { return _clients.size }
    );
};

connection.open();

File: python/backend.py
from twisted.internet import defer

from autobahn.twisted.wamp import Session, ApplicationRunner
from autobahn.wamp.types import RegisterOptions, PublishOptions
from autobahn import wamp

import random
import itertools


class Client(object):
    """
    Represents a single client's board state.

    'name' is their session-id and 'color' is a randomly-assigned
    color 3-tuple
    """
    def __init__(self, name, color):
        self.name = name
        self.color = color
        # a 16x16 grid of False values
        self._state = []
        for y in range(16):
            self._state.append([False] * 16)

    def click(self, x, y):
        self._state[x][y] = not self._state[x][y]


class Board(object):
    """
    The state of the game board.

    Each client has a bool for each pixel; if it's true, their color
    contributes to the total for that board square.
    """
    def __init__(self):
        # client_id -> Client instance
        self._clients = {}
        self._color_generator = itertools.cycle(
            [
                (255, 0, 0),
                (0, 255, 0),
                (0, 0, 255),
            ]
        )

    def click(self, client_id, x, y):
        self._clients[client_id].click(x, y)

    def client_add(self, client_id):
        color = next(self._color_generator)
        self._clients[client_id] = Client(client_id, color)

    def client_del(self, client_id):
        del self._clients[client_id]

    @wamp.register(u"meejah.get_client_count")
    def client_count(self):
        return len(self._clients)

    def _get_pixel(self, x, y):
        value = (0, 0, 0)
        for client in self._clients.values():
            if client._state[x][y]:
                value = (
                    max(value[0], client.color[0]),
                    max(value[1], client.color[1]),
                    max(value[2], client.color[2]),
                )
        return value

    @wamp.register(u"meejah.get_board")
    def as_json(self):
        # obviously, caching this would be a great idea...
        state = []
        for x in range(16):
            column = []
            for y in range(16):
                column.append(self._get_pixel(x, y))
            state.append(column)
        return state


class Game(Session):
    @defer.inlineCallbacks
    def on_join(self, details):
        self._board = Board()
        yield self.register(
            self._click, u"meejah.click",
            options=RegisterOptions(details_arg="details"),
        )
        yield self.register(self._board)
        yield self.subscribe(self)
        
    def on_leave(self, details):
        self.disconnect()

    @wamp.subscribe(u"wamp.session.on_leave")
    def _session_left(self, session):
        self._board.client_del(session)
        return self.publish(u"meejah.board_invalid")

    @wamp.subscribe(u"wamp.session.on_join")
    def _session_join(self, details):
        self._board.client_add(details['session'])
        return self.publish(u"meejah.board_invalid")

    @defer.inlineCallbacks
    def _click(self, x, y, details=None):
        """
        Public API, registerd as 'meejah.click'
        """
        self._board.click(details.caller, x, y)
        yield self.publish(
            u"meejah.board_update", x, y, self._board._get_pixel(x, y),
            options=PublishOptions(exclude_me=False),
        )


if __name__ == '__main__':
    runner = ApplicationRunner(u'ws://localhost:9999/ws', u'demo')
    # runner = ApplicationRunner(u'ws://yycjs.meejah.ca:9999/ws', u'demo')
    runner.run(Game)
